### ISSUE TITLE
Editor content conflict with content saved on disl

### ISSUE THREAD

**Issue by aeschli** [microsoft/vscode-hexeditor#276]
Testing https://github.com/microsoft/vscode/issues/119206

hexeditor-1.5.0.zip, MacOS, 
VSCode 1.55.0-insider, Commit: a6f6b4aa8e3a688d9a5cbcd5df14af511ce2bd8e, Date: 2021-03-23T05:13:33.822Z

- open a file in the hex editor
- copy the content of a character, and paste every second at different location

- at some point, a dialog will show up: `This file has changed on disk, but you have unsaved changes. Saving now will overwrite the file on disk with your changes.`

![hex](https://user-images.githubusercontent.com/6461412/112211866-9c0cf580-8c1c-11eb-9332-83f385a45982.gif)

**Comment by lramos15** [comment]
Do you have autosave on?

**Comment by aeschli** [comment]
Yes, `"files.autoSave": "afterDelay"`

**Comment by lramos15** [comment]
Hmm my guess is I should debounce the file watcher longer with autosave after a delay because what's happening here is the file watcher sees a change has been made and since it's after a little bit since your last save it assumes it's an external save.

**Comment by aeschli** [comment]
Maybe @bpasero can advise. I don't know how this is done with text files.

**Comment by bpasero** [comment]
@lramos15 I actually wanted to bring this up: with the work I am doing for file working copies for notebooks, I think custom editors could greatly benefit from this too provided we convince @mjbvz to add new API for simple file based custom editors. The idea is that you get all the benefits current text editors get with regards to save handling, dirty write prevention etc. without having to implement any of that. Even backups are being taken care of. All that is needed from an API point of view is `Buffer` based input and output, that's it.

**Comment by bpasero** [comment]
Check https://github.com/microsoft/vscode/issues/117873 and the video I posted in https://vscodeteam.slack.com/archives/G1C2N25T8/p1616408857050000

**Comment by lramos15** [comment]
Alright, that makes a lot of sense. Maybe we should set up a meeting to discuss the work that needs to be done there. I think these types of things are important as they help for a consistent experience across editor types

**Comment by bpasero** [comment]
@lramos15 yes totally, a sync between you, Matt and me would be good and maybe an issue to discuss this. The way I see it:
* we have a simple text based custom editor API already today
* we have a complex custom editor API where an extension has to implement everything
* with file working copy support coming along, we can have a third custom editor API which is simple and file based (for all binary cases like hex editor or image editor)

The API could be similar to what Jo is modeling in https://github.com/microsoft/vscode/issues/119089

**Comment by mjbvz** [comment]
@bpasero Using buffers goes back to some of our early sketches for the custom editor api. We ultimately decided against it for a few reason:

* Data transfer. It would require sending file buffers to and from the extension host 

* We don't have the equivalent of a `TextEdit` or `WorkspaceEdit` for binary files. 

* How do we implement undo/redo? Would VS Code maintain a stack of buffers? 

The hex editor is designed to open very large files, so it may not be a good fit for a buffer based API

**Comment by bpasero** [comment]
> Data transfer. It would require sending file buffers to and from the extension host

Yes that is true. Though this argument probably only applies to the hex editor, not the image editor from Daniel. 

> We don't have the equivalent of a TextEdit or WorkspaceEdit for binary files.

That is ok imho, we don't have that today for complex custom editors either, so nothing would really change.

> How do we implement undo/redo? Would VS Code maintain a stack of buffers?

This is maybe something to learn from notebooks how they will achieve this with their simple notebook API proposal.
